src/ts/components/TitleBar.ts

```diff
@@ -10,84 +12,105 @@ import { PlayerAPI } from 'bitmovin-player';
-
-    let checkMetadataTextAndUpdateVisibility = () => {
-      hasMetadataText = false;
-
-      // Iterate through metadata labels and check if at least one of them contains text
-      for (let component of this.getComponents()) {
-        if (component instanceof MetadataLabel) {
-          if (!component.isEmpty()) {
-            hasMetadataText = true;
-            break;
-          }
-        }
-      }
+    let hasVisibleText = true;
+
+    const checkTextAndUpdateVisibility = () => {
+      hasVisibleText = this.hasNonEmptyComponents(this.getComponents());
 
       if (this.isShown()) {
-        // Hide a visible titlebar if it does not contain any text and the hidden flag is set
-        if (config.keepHiddenWithoutMetadata && !hasMetadataText) {
+        if (config.keepHiddenWithoutMetadata && !hasVisibleText) {
           this.hide();
         }
       } else if (shouldBeShown) {
-        // Show a hidden titlebar if it should actually be shown
-        this.show();
+        if (!config.keepHiddenWithoutMetadata || hasVisibleText) {
+          this.show();
+        }
       }
     };
 
-    // Listen to text change events to update the hasMetadataText flag when the metadata dynamically changes
-    for (let component of this.getComponents()) {
-      if (component instanceof MetadataLabel) {
-        component.onTextChanged.subscribe(checkMetadataTextAndUpdateVisibility);
-      }
-    }
+    // Subscribe to any label-like component's text changes
+    this.subscribeToTextChanges(this.getComponents(), checkTextAndUpdateVisibility);
 
     uimanager.onControlsShow.subscribe(() => {
       shouldBeShown = true;
-      if (!(config.keepHiddenWithoutMetadata && !hasMetadataText)) {
+      if (!config.keepHiddenWithoutMetadata || hasVisibleText) {
         this.show();
       }
     });
+
     uimanager.onControlsHide.subscribe(() => {
       shouldBeShown = false;
       this.hide();
     });
 
-    // init
-    checkMetadataTextAndUpdateVisibility();
+    checkTextAndUpdateVisibility();
+  }
+
+  private hasNonEmptyComponents(components: Component<ComponentConfig>[]): boolean {
+    for (const component of components) {
+      if (hasIsEmpty(component) && !component.isEmpty()) {
+        return true;
+      }
+
+      if (component instanceof Container) {
+        if (this.hasNonEmptyComponents(component.getComponents())) {
+          return true;
+        }
+      }
+    }
+    return false;
   }
-}
\ No newline at end of file
+
+  private subscribeToTextChanges(components: Component<ComponentConfig>[], onChange: () => void): void {
+    for (const component of components) {
+      if (component instanceof Label) {
+        component.onTextChanged.subscribe(onChange);
+      }
+
+      if (component instanceof Container) {
+        this.subscribeToTextChanges(component.getComponents(), onChange);
+      }
+    }
+  }
+}
+
+function hasIsEmpty(obj: unknown): obj is { isEmpty: () => boolean } {
```

Comment:
👀 can you explain what this type is? (I'm not super into Typescript but this looks like black magic 😅)